### PR TITLE
Interrupt controller for e2k

### PR DESCRIPTION
**Pull request by alexkalmuk**
* Implement interrupts and exceptions handling
* ticker command now works on e2k

### CODE REVIEW

File: src/drivers/clock/e2k/e2k.c [hunk 1]
```diff
@@ -71,13 +72,14 @@
 
 static struct clock_source e2k_clock_source;
 static irq_return_t e2k_clock_handler(unsigned int irq_nr, void *dev_id) {
+	static int k = 0;
```
Comment (Deryugin) [inline src/drivers/clock/e2k/e2k.c]: Unused variable?

File: src/drivers/clock/e2k/e2k.c [hunk 2]
```diff
@@ -94,12 +96,17 @@ static int e2k_clock_config(struct time_dev_conf * conf) {
...
+#if 0
+	/* TODO Should be checked, because values are looking a bit weird. */
 	return e2k_read32((void*)E2K_RESET_COUNTER_LOW); /* Ignore high 32 bits */
+#endif
+	return 0;
 }
 
 static struct time_event_device e2k_clock_event = {
 	.config   = e2k_clock_config,
 	.event_hz = 1000,
```
Comment (Deryugin) [inline src/drivers/clock/e2k/e2k.c]: May be use `option number freq` instead?

File: src/arch/e2k/e2k_interrupt.c [hunk 3]
```diff
@@ -5,8 +5,74 @@
...
+
+	do {
+		++nr_tirs;
+		tirs[nr_tirs].hi = E2K_GET_DSREG(tir.hi);
+		tirs[nr_tirs].lo = E2K_GET_DSREG(tir.lo);
+		log_debug("TIR%d: lo=0x%lx, hi=0x%lx\n",
+			nr_tirs, tirs[nr_tirs].lo, tirs[nr_tirs].hi);
+	} while(GET_NR_TIRS(tirs[nr_tirs].hi));
+
+	assert(nr_tirs < MAX_TIRS);
```
Comment (Deryugin) [inline src/arch/e2k/e2k_interrupt.c]: I guess this `assert()` should be done before accessing `tirs[nr_tirs]`

File: src/arch/e2k/e2k_interrupt.c [hunk 4]
```diff
@@ -5,8 +5,74 @@
...
+#define GET_NR_TIRS(tir_hi)		((tir_hi >> 56) & 0xff)
+
+#define E2K_EXP_INTERRUPT  (1UL << 32)
+
+struct e2k_tir {
+	uint64_t lo;
+	uint64_t hi;
+};
+
+static void e2k_get_TIRs(struct e2k_tir *tirs) {
```
Comment (anton-bondarev) [inline src/arch/e2k/e2k_interrupt.c]: May be better e2k_get_TIRs -> e2k_get_tirs

File: src/arch/e2k/e2k_interrupt.c (hunk 3 above)
Comment (alexkalmuk) [reply src/arch/e2k/e2k_interrupt.c]: That's true, fixed.

File: src/drivers/clock/e2k/e2k.c (hunk 2 above)
Comment (alexkalmuk) [reply src/drivers/clock/e2k/e2k.c]: Fixed.

File: src/drivers/clock/e2k/e2k.c (hunk 1 above)
Comment (alexkalmuk) [reply src/drivers/clock/e2k/e2k.c]: Thanks, I removed this and the next commented out line.

**Pull request merged**